crates/parse/src/lib.rs

```diff
@@ -22,23 +22,31 @@ type ParseResult = Result<ast::Node, ParseError>;
 pub struct Parse<'a> {
     tokens: Peekable<Iter<'a, Token>>,
     symbol_table: &'a mut SymbolTable<Symbol>,
+    errs: Vec<ParseError>,
 }
 
 impl<'a> Parse<'a> {
     pub fn new(tokens: &'a [Token], symbol_table: &'a mut SymbolTable<Symbol>) -> Self {
-        Parse { tokens: tokens.iter().peekable(), symbol_table }
+        Parse { tokens: tokens.iter().peekable(), symbol_table, errs: Vec::new() }
     }
 
     // Parse each token using recursive descent
     //
     // StmtList ::= ( Stmt ';' )+ ;
-    pub fn parse(mut self) -> Result<Ast<ast::Node>, ParseError> {
+    pub fn parse(mut self) -> Result<Ast<ast::Node>, Vec<ParseError>> {
         let mut ast = Ast::new();
         while self.tokens.peek().is_some() {
-            let node = self.parse_stmt()?;
-            ast.add(node);
+            match self.parse_stmt() {
+                Ok(n) if self.errs.is_empty() => ast.add(n),
+                Err(e) => self.push_err(e),
+                _ => continue
```

Comment:
This is unreachable, right? If so change `continue` to `unreachable!("why this is unreachable...")`

Reply:
Actually, why not just skip the guard? Will it ever `parse_stmt()` ever return `Ok()` and `errs` be empty?

Reply:
After the parse has successfully recovered, every seccessfully parsed stmt will return Ok(n) with self.errs being non-empty. The last guard is not unreachable and cannot be skipped.